File: touchflow/focus_watcher.py
"""Отслеживание фокуса текстовых полей через AT-SPI (работает в GTK/Qt и др.)."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Callable

log = logging.getLogger(__name__)

try:
    import gi

    gi.require_version("Atspi", "2.0")
    from gi.repository import Atspi, GLib
except (ImportError, ValueError):
    Atspi = None  # type: ignore
    GLib = None  # type: ignore


@dataclass
class FocusInfo:
    app_id: str = ""
    window_class: str = ""
    role: str = ""
    is_text_entry: bool = False
    name: str = ""


TEXT_ROLES = {
    "entry",
    "password text",
    "text",
    "editable text",
    "terminal",
    "document text",
    "spin button",
    "combo box",
}


class FocusWatcher:
    def __init__(self, on_focus_change: Callable[[FocusInfo], None]):
        self._callback = on_focus_change
        self._listener_id: int | None = None
        self._available = Atspi is not None

    @property
    def available(self) -> bool:
        return self._available

    def start(self) -> bool:
        if not self._available:
            log.warning("AT-SPI unavailable — auto-show disabled")
            return False
        try:
            Atspi.init()
            self._listener_id = Atspi.EventListener.register_full(
                self._on_event,
                "object:state-change:focused",
            )
            log.info("AT-SPI focus watcher started")
            return True
        except Exception as e:
            log.error("Failed to start AT-SPI: %s", e)
            self._available = False
            return False

    def stop(self) -> None:
        if self._listener_id is not None and self._available:
            try:
                Atspi.EventListener.deregister(self._listener_id)
            except Exception:
                pass
            self._listener_id = None

    def _on_event(self, event) -> None:
        try:
            info = self._extract_focus(event.source)
            self._callback(info)
        except Exception as e:
            log.debug("Focus event error: %s", e)

    def _extract_focus(self, accessible) -> FocusInfo:
        info = FocusInfo()
        if accessible is None:
            return info

        try:
            role = accessible.get_role_name().lower()
            info.role = role
            info.name = accessible.get_name() or ""
            states = accessible.get_state_set()
            focused = states.contains(Atspi.StateType.FOCUSED)
            editable = states.contains(Atspi.StateType.EDITABLE)
            info.is_text_entry = focused and (role in TEXT_ROLES or editable)
        except Exception:
            pass

        try:
            app = accessible.get_application()
            if app:
                info.app_id = (app.get_name() or "").lower()
        except Exception:
            pass

        try:
            top = accessible
            for _ in range(20):
                parent = top.get_parent()
                if parent is None:
                    break
                top = parent
            info.window_class = (top.get_name() or "").lower()
        except Exception:
            pass

        return info

    def poll_current(self) -> FocusInfo:
        if not self._available:
            return FocusInfo()
        try:
            desktop = Atspi.get_desktop(0)
            focused = desktop.get_child_at_index(0)
            if focused:
                return self._extract_focus(focused)
        except Exception:
            pass
        return FocusInfo()

File: experimental/touchflow-cpp/src/focus_watcher.cpp
#include "focus_watcher.hpp"

#include <atspi/atspi.h>
#include <iostream>
#include <cstring>

namespace touchflow {

static bool is_text_role(const char* role) {
    if (!role) return false;
    static const char* roles[] = {
        "entry", "password text", "text", "editable text",
        "terminal", "document text", "spin button", "combo box", nullptr
    };
    for (const char** r = roles; *r; ++r)
        if (std::strcmp(role, *r) == 0) return true;
    return false;
}

FocusWatcher::FocusWatcher(Callback cb) : callback_(std::move(cb)) {}

FocusWatcher::~FocusWatcher() { stop(); }

void FocusWatcher::deliver(_AtspiAccessible* focused) {
    if (!callback_ || !focused) return;
    auto* acc = reinterpret_cast<AtspiAccessible*>(focused);
    FocusInfo info;
    gchar* role = atspi_accessible_get_role_name(acc, nullptr);
    info.role = role ? role : "";
    g_free(role);
    info.is_text_entry = is_text_role(info.role.c_str());

    AtspiAccessible* app = atspi_accessible_get_application(acc, nullptr);
    if (app) {
        gchar* name = atspi_accessible_get_name(app, nullptr);
        info.app_id = name ? name : "";
        g_free(name);
        g_object_unref(app);
    }
    callback_(info);
}

void atspi_focus_cb(_AtspiEvent* raw, void* data) {
    auto* event = reinterpret_cast<AtspiEvent*>(raw);
    auto* self = static_cast<FocusWatcher*>(data);
    if (!self || !event || !event->source) return;
    if (event->detail1 != 1) return;
    self->deliver(event->source);
}

bool FocusWatcher::start() {
    GError* err = nullptr;
    if (!atspi_init()) {
        std::cerr << "touchflow-cpp: AT-SPI init failed\n";
        return false;
    }
    auto* listener = atspi_event_listener_new(atspi_focus_cb, this, nullptr);
    if (!listener) return false;
    if (!atspi_event_listener_register(listener, "object:state-change:focused", &err)) {
        if (err) {
            std::cerr << "touchflow-cpp: AT-SPI register: " << err->message << "\n";
            g_error_free(err);
        }
        g_object_unref(listener);
        return false;
    }
    listener_ = listener;
    available_ = true;
    return true;
}

void FocusWatcher::stop() {
    if (listener_) {
        GError* err = nullptr;
        atspi_event_listener_deregister(
            static_cast<AtspiEventListener*>(listener_),
            "object:state-change:focused", &err);
        if (err) g_error_free(err);
        g_object_unref(listener_);
        listener_ = nullptr;
    }
}

}  // namespace touchflow
